File: 13_Command_Line_Arguments/java/CliArguments.java
import java.util.ArrayList;
import java.util.List;

public class CliArguments {

    static class Config {
        String name = "";
        int count = 1;
        boolean verbose = false;
        boolean help = false;
        List<String> files = new ArrayList<>();
    }

    static void printUsage() {
        System.out.println("Usage: java CliArguments [options] [files...]");
        System.out.println("Options:");
        System.out.println("  -n, --name <name>   Set the name");
        System.out.println("  -c, --count <num>   Set repeat count (default: 1)");
        System.out.println("  -v, --verbose       Enable verbose output");
        System.out.println("  -h, --help          Show this help message");
    }

    static Config parseArgs(String[] args) {
        Config config = new Config();

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-h": case "--help":
                    config.help = true;
                    break;
                case "-v": case "--verbose":
                    config.verbose = true;
                    break;
                case "-n": case "--name":
                    if (i + 1 < args.length) {
                        config.name = args[++i];
                    } else {
                        System.err.println("Error: " + args[i] + " requires a value");
                    }
                    break;
                case "-c": case "--count":
                    if (i + 1 < args.length) {
                        try {
                            config.count = Integer.parseInt(args[++i]);
                            if (config.count < 1) {
                                System.err.println("Error: count must be positive");
                                config.count = 1;
                            }
                        } catch (NumberFormatException e) {
                            System.err.println("Error: Invalid number for --count");
                        }
                    } else {
                        System.err.println("Error: " + args[i] + " requires a value");
                    }
                    break;
                default:
                    if (args[i].startsWith("-")) {
                        System.err.println("Warning: Unknown option '" + args[i] + "'");
                    } else {
                        config.files.add(args[i]);
                    }
                    break;
            }
        }
        return config;
    }

    public static void main(String[] args) {
        System.out.println("=== Command Line Arguments Demo ===");

        System.out.println("\n--- Raw Arguments ---");
        System.out.println("args.length = " + args.length);
        for (int i = 0; i < args.length; i++) {
            System.out.println("  args[" + i + "] = \"" + args[i] + "\"");
        }

        if (args.length == 0) {
            System.out.println("\nNo arguments provided. Running with defaults.");
            System.out.println("Try: java CliArguments --name Alice -c 3 -v file1.txt file2.txt");
        }

        Config config = parseArgs(args);

        if (config.help) {
            System.out.println("\n--- Help ---");
            printUsage();
            return;
        }

        System.out.println("\n--- Parsed Configuration ---");
        System.out.println("  Name:    " + (config.name.isEmpty() ? "(not set)" : config.name));
        System.out.println("  Count:   " + config.count);
        System.out.println("  Verbose: " + (config.verbose ? "yes" : "no"));
        System.out.print("  Files:   ");
        if (config.files.isEmpty()) {
            System.out.println("(none)");
        } else {
            System.out.println(String.join(", ", config.files));
        }

        if (!config.name.isEmpty()) {
            System.out.println("\n--- Output ---");
            for (int i = 0; i < config.count; i++) {
                System.out.println("  Hello, " + config.name + "!");
            }
        }

        if (config.verbose && !config.files.isEmpty()) {
            System.out.println("\n--- File Details (verbose) ---");
            for (String f : config.files) {
                System.out.println("  File: " + f + " (" + f.length() + " chars in name)");
            }
        }
    }
}

File: 13_Command_Line_Arguments/cpp/cli_arguments.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

struct Config {
    string name = "";
    int count = 1;
    bool verbose = false;
    bool help = false;
    vector<string> files;
};

void printUsage(const string& program) {
    cout << "Usage: " << program << " [options] [files...]" << endl;
    cout << "Options:" << endl;
    cout << "  -n, --name <name>   Set the name" << endl;
    cout << "  -c, --count <num>   Set repeat count (default: 1)" << endl;
    cout << "  -v, --verbose       Enable verbose output" << endl;
    cout << "  -h, --help          Show this help message" << endl;
}

Config parseArgs(int argc, char* argv[]) {
    Config config;

    for (int i = 1; i < argc; i++) {
        string arg = argv[i];

        if (arg == "-h" || arg == "--help") {
            config.help = true;
        } else if (arg == "-v" || arg == "--verbose") {
            config.verbose = true;
        } else if (arg == "-n" || arg == "--name") {
            if (i + 1 < argc) {
                config.name = argv[++i];
            } else {
                cerr << "Error: " << arg << " requires a value" << endl;
            }
        } else if (arg == "-c" || arg == "--count") {
            if (i + 1 < argc) {
                try {
                    config.count = stoi(argv[++i]);
                } catch (...) {
                    cerr << "Error: Invalid number for --count" << endl;
                }
            } else {
                cerr << "Error: " << arg << " requires a value" << endl;
            }
        } else if (arg[0] == '-') {
            cerr << "Warning: Unknown option '" << arg << "'" << endl;
        } else {
            config.files.push_back(arg);
        }
    }
    return config;
}

int main(int argc, char* argv[]) {
    cout << "=== Command Line Arguments Demo ===" << endl;

    cout << "\n--- Raw Arguments ---" << endl;
    cout << "argc = " << argc << endl;
    for (int i = 0; i < argc; i++) {
        cout << "  argv[" << i << "] = \"" << argv[i] << "\"" << endl;
    }

    Config config = parseArgs(argc, argv);

    if (config.help) {
        cout << "\n--- Help ---" << endl;
        printUsage(argv[0]);
        return 0;
    }

    cout << "\n--- Parsed Configuration ---" << endl;
    cout << "  Name:    " << (config.name.empty() ? "(not set)" : config.name) << endl;
    cout << "  Count:   " << config.count << endl;
    cout << "  Verbose: " << (config.verbose ? "yes" : "no") << endl;
    cout << "  Files:   ";
    if (config.files.empty()) {
        cout << "(none)";
    } else {
        for (size_t i = 0; i < config.files.size(); i++) {
            if (i > 0) cout << ", ";
            cout << config.files[i];
        }
    }
    cout << endl;

    if (!config.name.empty()) {
        cout << "\n--- Output ---" << endl;
        for (int i = 0; i < config.count; i++) {
            cout << "  Hello, " << config.name << "!" << endl;
        }
    }

    if (config.verbose && !config.files.empty()) {
        cout << "\n--- File Details (verbose) ---" << endl;
        for (const auto& f : config.files) {
            cout << "  File: " << f << " (" << f.length() << " chars in name)" << endl;
        }
    }

    cout << "\nTry: " << argv[0] << " --name Alice -c 3 -v file1.txt file2.txt" << endl;
    return 0;
}
